### PR TITLE
Fix and add more padding mode support for Conv

### LINKED ISSUE #29712: conv circular padding is not working
## 🐛 Bug
Not only it is not documented #29668 , it does nothing
```py
In [1]: import torch

In [2]: torch.__version__
Out[2]: '1.4.0a0+f94b6ce'

In [3]: x = torch.randn(1, 3, 5, 5)
   ...:
   ...: c = torch.nn.Conv2d(3, 3, 3, padding=1, padding_mode='circular')
   ...:
   ...: c(x).shape
Out[3]: torch.Size([1, 3, 3, 3])

In [4]:

In [4]: x = torch.randn(1, 3, 5, 5)
   ...:
   ...: c = torch.nn.Conv2d(3, 3, 3, padding=1)
   ...:
   ...: c(x).shape
Out[4]: torch.Size([1, 3, 5, 5])
```

cc @ezyang

### PR DESCRIPTION
**Pull request by ssnl**
Fix #29712 #29668 , add arg checking, doc, and support for reflection and replication padding modes.

### CODE REVIEW

**Comment by kostmo** [comment]
## :pill: CircleCI build failures summary and remediations
As of commit fead294b:


**None of the build failures appear to be your fault**.

* **1/1** [broken upstream](https://dr.pytorch.org/master-timeline.html?count=1&sha1=9e9bfbfd8d893cdf8205db67f220551214787d7f&should_suppress_scheduled_builds=true&should_suppress_fully_successful_columns=true&max_columns_suppress_successful=35) at merge base 9e9bfbfd since Jan 02
  <details>
  <summary>
  You may want to rebase on the <code>viable/strict</code> branch (expand for instructions)
  </summary>
  
  If your commit is newer than `viable/strict`, you can try basing on an older, stable commit:
  ```
  git fetch origin viable/strict
  git rebase --onto viable/strict $(git merge-base origin/master HEAD)
  ```
  If your commit is older than `viable/strict`:
  ```
  git fetch origin viable/strict
  git rebase viable/strict
  ```
  Check out the [recency history](https://dr.pytorch.org/master-viable-commits.html) of this "viable master" tracking branch.
  
  </details>
  


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=fead294b31bbcbd054bfb8305204d98932d2d007).

### :construction: 1 upstream failure recognized by patterns:
These builds matched patterns, but were probably **caused by upstream breakages**:
* [pytorch_linux_backward_compatibility_check_test](https://dr.pytorch.org/build-details.html?build_id=79380126) from Jan 02



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

File: torch/nn/modules/conv.py [hunk 1]
```diff
@@ -32,7 +32,12 @@ def __init__(self, in_channels, out_channels, kernel_size, stride,
         self.transposed = transposed
         self.output_padding = output_padding
         self.groups = groups
+        if padding_mode not in {'zeros', 'reflect', 'replicate', 'circular'}:
+            raise ValueError(
+                "padding_mode must be one of {{'zeros', 'reflect', 'replicate', 'circular'}}, "
```
Comment (ezyang) [inline torch/nn/modules/conv.py]: nit: it would be better to assign the set of valid padding modes to a variable so you don't have to repeat it

**Review by ezyang — APPROVED** [review]

File: torch/nn/modules/conv.py (hunk 1 above)
Comment (SsnL) [reply torch/nn/modules/conv.py]: good suggestion. fixed!

**Comment by ezyang** [comment]
Sorry, it looks like this is breaking BC-compat with some people who are using some internal identifiers:

```
AttributeError: module 'torch.nn.modules.conv' has no attribute '_ConvTransposeMixin'
```

Should we maybe just keep the old names around?

**Comment by SsnL** [comment]
@ezyang Oh... but this isn't a mixin anymore. It would be weird if it is still called a mixin. How is it used? If people use it in multiple inheritance with `_ConvNd`, just keeping the old name won't work...

**Comment by ezyang** [comment]
They're defining convolution like modules. The code looks something like:

```
class QuantConvTranspose2d(torch.nn.modules.conv._ConvTransposeMixin, torch.nn.modules.conv._ConvNd):
  ...
```

I think it was just cargo culted off of our existing convolution clasess.

**Comment by SsnL** [comment]
@ezyang I can add an alias `_ConvTransposeMixin  = _ConvTransposeNd`. But it will be quite confusing... If not too hard and there aren't too many, could you change those to inheriting from just a single `_ConvTransposeNd` in the diff? Thanks!!

**Comment by ezyang** [comment]
I could fix it (I don't think there are many definitions). But I'm mentioning it here because often these internal issues are a canary for uses out in the wild. I did a quick GitHub code search and here is another occurrence: https://github.com/taokong/FoveaBox/blob/47ed4326593beb50cf2a691c0ee8e0bdedc7ce73/mmdet/utils/flops_counter.py

**Comment by ezyang** [comment]
Another one: https://github.com/stormraiser/gan-weightnorm-resnet/blob/18695b138a86a255f9c38d22be2e80bad1087d6f/modules/WeightNormalizedConv.py

**Comment by SsnL** [comment]
@ezyang I added an alias. Hopefully this resolves the BC issue.

**Comment by ezyang** [comment]
LGTM

**Comment by zou3519** [comment]
I removed the BC-breaking label because it looks like this PR doesn't break BC but instead introduces a deprecation, let me know if this is wrong.

**Pull request closed without merging**